### PR TITLE
Painless: Clean up FunctionRef

### PR DESCRIPTION
**Pull request by jdconrad**
This change consolidates all the logic for generating a FunctionReference (renamed from FunctionRef) from several arbitrary constructors to a single static function that is used at both compile-time and run-time.  This increases long-term maintainability as it is much easier to follow when and how a function reference is being generated.  It moves most of the duplicated logic out of the ECapturingFuncRef, EFuncRef and ELambda nodes and Def as well.

### CODE REVIEW

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/FunctionReference.java [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+import java.util.Map;
+import java.util.Objects;
+
+import static org.elasticsearch.painless.WriterConstants.CLASS_NAME;
+import static org.objectweb.asm.Opcodes.H_INVOKEINTERFACE;
+import static org.objectweb.asm.Opcodes.H_INVOKESTATIC;
+import static org.objectweb.asm.Opcodes.H_INVOKEVIRTUAL;
+import static org.objectweb.asm.Opcodes.H_NEWINVOKESPECIAL;
+
+public class FunctionReference {
```
Comment (rjernst) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/FunctionReference.java]: It looks like the top level javadoc was lost in the rename?

**Review by rjernst — CHANGES_REQUESTED** [review]
I think the original name FunctionRef actually fit better into existing names within the codebase, eg we have BytesRef. The rename is also very difficult to review, as it is unclear what little changes there are within the file since it looks all new. Can you change the name back, and if the rename is that important, do it in a followup PR by itself?

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/Def.java [hunk 2]
```diff
@@ -331,57 +331,34 @@ static MethodHandle lookupMethod(PainlessLookup painlessLookup, Map<String, Loca
       */
     static MethodHandle lookupReference(PainlessLookup painlessLookup, Map<String, LocalMethod> localMethods,
             MethodHandles.Lookup methodHandlesLookup, String interfaceClass, Class<?> receiverClass, String name) throws Throwable {
-         Class<?> interfaceType = painlessLookup.canonicalTypeNameToType(interfaceClass);
-         PainlessMethod interfaceMethod = painlessLookup.lookupPainlessClass(interfaceType).functionalMethod;
-         if (interfaceMethod == null) {
+            Class<?> interfaceType = painlessLookup.canonicalTypeNameToType(interfaceClass);
```
[the remark this answers is not in the document]
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/Def.java]: Fixed.  For some reason the formatting in this file has a lot of issues.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/FunctionReference.java (hunk 1 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/FunctionReference.java]: Added a new one.  The old one wasn't descriptive.

**Comment by jdconrad** [comment]
@rjernst Thanks for the review!  Ready for the next round.

Renamed this back to FunctionRef; however, while we do have classes like BytesRef outside of Painless, we've been heading toward longer more descriptive file names in Painless, and I would like to change this back in a different PR.  The file names in Painless are all over the place right now and I would like a stronger consistency moving forward as things are worked on.

**Review by rjernst — APPROVED** [review]
LGTM. Thanks!

**Comment by jdconrad** [comment]
@rjernst Thanks again for the review!  Will commit as soon as CI passes.

**Pull request merged**